[feat] Add flaky test verification workflow

sfc-gh-nbellante opened the pull request:
## Describe your changes

Added a new GitHub workflow for flaky test verification that allows running E2E tests multiple times to identify intermittent failures. The workflow can be triggered in two ways:

1. By adding the `flaky-verify` label to a PR (This one I have not been able to test because as far as I can tell, it only works when the workflow already exists in develop? If someone knows better, please correct me here.)
2. By manually triggering the workflow with a PR number and optional run count (This one is working: https://github.com/streamlit/streamlit/actions/workflows/flaky-test-verification.yml)

Also enhanced the existing `playwright-changed-files.yml` workflow to:
- Support being called from other workflows via `workflow_call`
- Accept custom git refs and base refs for comparison
- Support matrix runs with unique concurrency groups
- Improve changed file detection by comparing against a specified base ref
- Add better artifact naming for matrix runs

## Testing Plan

The flaky test verification workflow has been thoroughly tested using manual triggers:

  #### Test with 2 runs
 `gh workflow run flaky-test-verification.yml --ref 11-04-flaky_test_detector -f pr_number=12922 -f runs=2`

  #### Test with 5 runs
  `gh workflow run flaky-test-verification.yml --ref 11-04-flaky_test_detector -f pr_number=12922 -f runs=5`

  Verified functionality:
  - ✅ Workflow successfully detects PR base branch/SHA (supports Graphite stacks)
  - ✅ Triggers multiple concurrent runs of E2E tests
  - ✅ Correctly identifies changed test files
  - ✅ Runs changed tests against the PR merge ref
  - ✅ Works with stacked PRs (uses base SHA instead of branch name)

>[!NOTE]
> This originally used `pull_request_target`, but we switched it to use `pull_request` for security reasons and to closer align with our other workflows
>

  🧪 How to Test After Merge

  Once this PR is merged to develop:

  1. Go to any PR with E2E test changes (e.g., PR #12922, #12923, or #12927)
  2. Add the flaky-verify label
  3. The workflow should automatically trigger and run the tests N times (default: 50)
  4. Check the "Actions" tab to see the workflow run

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-nbellante committed:
flaky test detector

sfc-gh-nbellante commented:
* **#12929** <a href="https://app.graphite.com/github/pr/streamlit/streamlit/12929?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 2  dependent PRs ([#12922](https://github.com/streamlit/streamlit/pull/12922) <a href="https://app.graphite.com/github/pr/streamlit/streamlit/12922?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#12923](https://github.com/streamlit/streamlit/pull/12923) <a href="https://app.graphite.com/github/pr/streamlit/streamlit/12923?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>) 👈 <a href="https://app.graphite.com/github/pr/streamlit/streamlit/12929?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12921** <a href="https://app.graphite.com/github/pr/streamlit/streamlit/12921?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sfc-gh-nbellante committed:
Fix workflow syntax: move uses to job level

sfc-gh-nbellante committed:
Use workflow_dispatch to trigger multiple test runs

sfc-gh-nbellante committed:
Fix env variable reference in if condition

sfc-gh-nbellante committed:
Add actions:write permission to trigger workflows

sfc-gh-nbellante committed:
Trigger workflow on current branch to find workflow_dispatch

sfc-gh-nbellante committed:
Use PR base branch for git diff comparison to support Graphite stacks

sfc-gh-nbellante committed:
Use base SHA instead of base ref for Graphite stack support

sfc-gh-nbellante committed:
Remove AST-based test detection, just run entire changed test files

sfc-gh-nbellante committed:
Add pull_request trigger for testing label-based activation

sfc-gh-nbellante committed:
Revert pull_request trigger - doesn't help with testing

sfc-gh-nbellante committed:
Use workflow_call instead of workflow_dispatch to wait for test completion

sfc-gh-nbellante committed:
Add base_ref input to workflow_call trigger

sfc-gh-nbellante committed:
Add run_id to concurrency group to allow parallel flaky verification runs

sfc-gh-nbellante committed:
Remove concurrency control to allow parallel matrix execution

The workflow-level concurrency group was preventing matrix jobs from running
in parallel. Since all matrix iterations share the same github.ref and
github.run_id, they were being queued and cancelled according to GitHub's
concurrency rules.

Removing the concurrency block entirely allows:
- Matrix jobs to run in parallel when invoked via workflow_call
- Pull request workflows still benefit from GitHub's built-in PR-level concurrency

sfc-gh-nbellante committed:
Fix concurrency to allow parallel matrix runs while preserving PR cancellation

Pass iteration number as input to playwright-changed-files workflow and include
it in the concurrency group. This makes each matrix iteration unique while still
allowing duplicate PR runs to cancel each other:

- Matrix iteration 1: group ends with '-1-playwright-changed-files'
- Matrix iteration 2: group ends with '-2-playwright-changed-files'
- Normal PR runs: group ends with '--playwright-changed-files' (iteration empty)

This fixes the issue where all matrix jobs were in the same concurrency group
and being cancelled, while maintaining the desired behavior for normal PR runs.

sfc-gh-nbellante committed:
Fix artifact name collisions in parallel matrix runs

Include iteration number in artifact names to prevent 409 Conflict errors
when multiple matrix jobs try to upload artifacts with the same name.

Artifact naming:
- Normal runs: playwright_test_results_{sha}
- Matrix runs: playwright_test_results_{sha}_iter{N}

sfc-gh-nbellante committed:
Fix fetch-depth for PR head checkout to enable merge base detection

The workflow was failing with "no merge base" error because we were using
fetch-depth: 2 when checking out PR heads, which is too shallow to find
a merge base with the develop branch.

Changed fetch-depth from 2 to 0 (full history) for PR head checkouts to
match the behavior of ref input checkouts. This allows git diff to properly
find the merge base when comparing against the base branch.

Fixes the error: "fatal: develop...HEAD: no merge base"

sfc-gh-nbellante committed:
Address PR feedback: improve flaky test verification workflow

Made the following improvements based on Graphite Agent feedback:

1. Added pull-requests: read permission
   - Required for gh api calls to fetch PR metadata
   - Prevents 403 errors when manually triggering workflow

2. Fixed variable naming consistency
   - Changed base-ref to base_ref throughout (lines 34, 82)
   - Ensures proper data passing between jobs

3. Added comprehensive PR number validation
   - Validates workflow_dispatch has PR number input
   - Fails fast with clear error messages
   - Handles pull_request_target events properly
   - Catches unsupported event types

These changes improve workflow robustness and user experience by
providing clear error messages and ensuring proper permissions.

sfc-gh-nbellante committed:
Restore 'Use output' debug step in playwright-changed-files workflow

This debug step was accidentally removed when cleaning up the AST-based
test detection code. It's useful for debugging the run_tests output value.

sfc-gh-nbellante committed:
Fix critical bug: validate ref input for workflow_call events

The fallback checkout step (line 70-77) was using github.head_ref, which
is only available for pull_request events, not workflow_call events.

This would cause the workflow to checkout the wrong ref (or fail) if
called via workflow_call without a ref input.

Changes:
1. Added validation step that fails fast with clear error if ref is
   empty when called via workflow_call
2. Updated fallback checkout condition to only run for pull_request
   events (where github.head_ref is actually available)

This prevents silent failures and ensures the workflow fails early
with a clear error message if misconfigured.

sfc-gh-bnisco commented on .github/workflows/flaky-test-verification.yml:
```diff
@@ -0,0 +1,92 @@
+  pull-requests: read
+
+env:
+  FLAKY_VERIFY_LABEL: flaky-verify
```
**suggestion:** I don't see this env var being used anywhere, so it should be removed.

sfc-gh-nbellante replied:
Yeah, I thought I was going to need to use it more, but didn't use it enough so let me remove it.

sfc-gh-bnisco commented on .github/workflows/flaky-test-verification.yml:
```diff
@@ -0,0 +1,92 @@
+    if: |
+      github.event_name == 'workflow_dispatch' ||
+      (github.event_name == 'pull_request_target' && github.event.action == 'labeled' && github.event.label.name == 'flaky-verify')
+    runs-on: ubuntu-latest
```
**note:** This is technically a smaller machine than what the usual e2e test suite runs on, which is either `ubuntu-latest-8-cores` or `ubuntu-latest-32-cores` depending on the job type. I don't know if it matters in practice, but might be worth leaving a comment.

sfc-gh-bnisco commented on .github/workflows/playwright-changed-files.yml:
```diff
@@ -134,21 +192,24 @@ jobs:
+      - name: Run changed playwright tests
+        if: steps.check_changed_files.outputs.run_tests == 'true'
+        env:
+          FILES: ${{ steps.changed-files.outputs.all_changed_files }}
```
**question:** Is this env var necessary? It seems like we only use it in the line below, which seems like an unnecessary change to this file.

sfc-gh-bnisco reviewed (commented):
Very cool stuff!

Meta-question: why is this in a Graphite stack to `nico/add-worktrees-script`? This seems like it should be merged into `develop` directly?

sfc-gh-nbellante commented:
> Meta-question: why is this in a Graphite stack to `nico/add-worktrees-script`? This seems like it should be merged into `develop` directly?

Meta-answer: I used the worktree creation script in development of this feature 🤣

sfc-gh-nbellante committed:
Remove unused FLAKY_VERIFY_LABEL environment variable

Address feedback from Bob (sfc-gh-bnisco): The FLAKY_VERIFY_LABEL
env var was defined but never used. The label check in the workflow
condition hardcodes 'flaky-verify' directly, so the env var was
unnecessary dead code.

sfc-gh-nbellante committed:
Address Bob's feedback: documentation and cleanup

- Add comment explaining machine size choice in flaky-test-verification.yml
  (ubuntu-latest is sufficient for lightweight coordination job)
- Remove unnecessary FILES env var in playwright-changed-files.yml
  (inline expression directly where used for simplicity)

Addresses feedback from Bob (sfc-gh-bnisco) on PR #12929.

sfc-gh-nbellante commented on .github/workflows/playwright-changed-files.yml:
```diff
@@ -19,18 +54,41 @@ jobs:
+          # If it's a SHA, git already has it (we used fetch-depth: 0)
+
+          # Get all changed files comparing against base
+          git_output=$(git diff --name-only "${BASE_REF}...HEAD")
```
The triple-dot syntax is intentional here. We want to see the changes from the merge base to HEAD, which is exactly what `...` provides. Even though we're passing a SHA and working in a detached HEAD state (typical in GitHub Actions), the triple-dot syntax correctly finds the merge base and shows the changes introduced by the PR branch.

Using `..` (double-dot) would show all commits reachable from HEAD but not from BASE_REF, which could include changes that aren't part of this PR if there have been other commits to the base branch. The triple-dot syntax gives us the PR's actual changes, which is what we want for detecting which test files changed.

sfc-gh-nbellante committed:
Address PR feedback: improve workflow robustness

Fixed four issues identified in code review:

1. SHA detection logic: Use regex pattern to properly distinguish between
   commit SHAs and branch names. Previous logic failed to fetch branches
   with slashes (e.g., 'nico/feature') and incorrectly tried to fetch SHAs.

2. PR number validation: Add input validation to ensure pr_number is a
   positive integer before using in API calls. Provides better error
   messages and follows defensive coding practices.

3. Concurrency group formatting: Use conditional expression to avoid
   double-dash when iteration is empty, improving readability.

4. Base SHA error handling: Add validation to fail fast with clear error
   message if GitHub API call fails or returns empty data, preventing
   confusing downstream git errors.

sfc-gh-bnisco commented on .github/workflows/flaky-test-verification.yml:
```diff
@@ -0,0 +1,102 @@
+
+on:
+  pull_request_target:
+    types: [labeled]
```
**question:** Is `pull_request_target` the correct event to use here?

Currently no other GH Workflow in our repo uses `pull_request_target`, they all use `pull_request`. For example, the closest similar job to this one [is the autofix workflow](https://github.com/streamlit/streamlit/blob/develop/.github/workflows/autofix.yml#L5), which also uses `pull_request`.

I'm also weary of the warning that is stated in [the docs](https://docs.github.com/en/actions/reference/workflows-and-actions/events-that-trigger-workflows#pull_request_target):

> Running untrusted code on the pull_request_target trigger may lead to security vulnerabilities.

sfc-gh-nbellante replied:
We could always try it with `pull_request` and see what happens. 🤷‍♂️ 

The reason it got added was to have proper auth credentials for things like: 
1. `gh api repos/streamlit/streamlit/pulls/$pr_number`
2. `refs/pull/${{ needs.verify.outputs.pr-number }}/merge`
3. and triggering other workflows

sfc-gh-nbellante replied:
Admittedly, I haven't found a way to test the label-run workflow yet... I figured I'd just test it in prod 😅

sfc-gh-bnisco replied:
I think we need to be cautious of `pull_request_target` and we should scope this back to just `pull_request`. While I understand it may limit something like `gh api`, it seems like the main intent is to enable the PR label usecase. Let's optimize for getting that to work first and minimize risk of this change introducing anything net-new to our GH Workflows.

sfc-gh-nbellante committed:
Change pull_request_target to pull_request trigger

Address security concern raised by Bob in PR review. The workflow
only performs read operations (fetching PR metadata via gh api) and
doesn't need the elevated permissions that pull_request_target provides.

Using pull_request is more appropriate and follows the pattern used
by all other workflows in the repo (e.g., autofix.yml).

Changes:
- Switch trigger from pull_request_target to pull_request
- Update all event name checks in conditions and shell scripts

sfc-gh-bnisco reviewed (approved):
Thanks for the iterations. Please update the PR description to reflect the changes made to `pull_request` type
